File: src/jyotish/varshaphala.ts
import { Body, Equator, Horizon, MakeTime, Observer } from 'astronomy-engine';
import { computeRashiChart } from './charts';
import { computeShadbala } from './shadbala';
import { computeLagna } from './lagna';
import { RASHI_LORD } from './matchingTables';
import {
  ALL_SAHAM_NAMES, SAHAM_FORMULAS,
  type SahamFormula, type SahamName, type SahamOperand,
} from './sahamsTables';
import { getSiderealSunLongitude } from '../astronomy/sun';
import { resolveMasaName } from '../i18n/resolver';
import { normalize360 } from '../utils/angle';
import { validateDate, validateLocation } from '../utils/validation';
import { PanchangError } from '../types/errors';
import type { AyanamsaType, BirthChartOptions, Language } from '../types/options';
import type { GeoLocation } from '../types/location';
import type {
  BhavaChart, BirthChart, GrahaName, LagnaInfo, PlanetPlacement,
} from '../types/jyotish';

// ── Types ──────────────────────────────────────────────

/**
 * One Saham's resolved position in the varsha (annual) chart.
 *
 * A Saham is computed as `(X − Y + Z) mod 360°` (with X/Y swap on night
 * birth for the Sahams flagged in {@link SAHAM_FORMULAS}). The `longitude`
 * is the sidereal ecliptic point; `rashi` and `house` derive from that.
 */
export interface SahamPosition {
  /** Sidereal longitude of the Saham, [0, 360). */
  longitude: number;
  /** Zodiac sign the Saham occupies, 0..11 (Mesha … Meena). */
  rashi: number;
  /** Localized rashi name (`en` or `hi`) — same locale as the varsha chart. */
  rashiName: string;
  /** House (1..12) the Saham occupies, whole-sign from the varsha lagna. */
  house: number;
}

/**
 * Muntha — the year-by-year sensitive point that advances one rashi per
 * birthday. Anchored at the **natal** lagna's rashi at age 0; advances
 * by `yearAge` rashis (mod 12).
 */
export interface MunthaInfo {
  /** Rashi index (0..11) the Muntha occupies in the varsha chart. */
  rashi: number;
  /** Lord of the Muntha rashi (one of the 7 visible grahas). */
  lord: Exclude<GrahaName, 'Rahu' | 'Ketu'>;
  /** House (1..12) the Muntha occupies, whole-sign from the varsha lagna. */
  house: number;
}

/**
 * Full Varshaphala (Tajik annual) chart for a given solar-return year.
 *
 * The chart is cast at the **solar-return instant** — the UTC moment in
 * year-N when the sidereal Sun returns to its natal longitude (within
 * 0.0001° / ~9 arcseconds). Subsequent Tajik analytics layer on this:
 *
 *   - `varshaLagna` — sidereal ascendant at the solar-return instant.
 *   - `muntha` — the natal-lagna-anchored sensitive point advanced by
 *     `yearAge` rashis.
 *   - `yearLord` — the strongest of four Tajik year-lord candidates,
 *     selected by total Shadbala from the varsha chart.
 *   - `sahams` — the 27 core-set Sahams (Punya, Vidya, …, Tapas).
 *   - `planets` — full 9-graha placement at the varsha instant
 *     (re-uses `computeRashiChart`).
 *   - `bhava` — house cusps under the configured house system.
 */
export interface VarshaphalaChart {
  /** UTC moment of the sidereal solar return for `yearAge`. */
  solarReturnInstant: Date;
  /** Sidereal ascendant at the solar-return instant. */
  varshaLagna: LagnaInfo;
  /** Muntha — annual sensitive point. */
  muntha: MunthaInfo;
  /** Year lord (Varsha Pati) by max Shadbala among the four classical candidates. */
  yearLord: Exclude<GrahaName, 'Rahu' | 'Ketu'>;
  /** All 27 core Sahams keyed by name. */
  sahams: Record<SahamName, SahamPosition>;
  /** True if the solar-return instant occurred during day (Sun above horizon). */
  isDayBirth: boolean;
  /** 9-graha placements in the varsha chart. */
  planets: PlanetPlacement[];
  /** House cusps for the varsha chart under the configured house system. */
  bhava: BhavaChart;
}

// ── Constants ──────────────────────────────────────────

/**
 * Sidereal year (Sun returning to the same sidereal longitude) in days.
 * Used as the per-year stride for the solar-return initial guess.
 *
 * 365.25636 days = 365 d 6 h 9 m 9.7 s. Within 1 day of the natal
 * anniversary even after 100 years, so the Newton refinement below
 * converges in ~3–5 iterations.
 */
const SIDEREAL_YEAR_DAYS = 365.25636;

/** Sun's mean sidereal motion, degrees per day. Used to convert delta to time step. */
const SUN_DEG_PER_DAY = 360 / SIDEREAL_YEAR_DAYS; // ≈ 0.98561

type VisibleGraha = Exclude<GrahaName, 'Rahu' | 'Ketu'>;

const VISIBLE_GRAHAS_BY_INDEX: readonly VisibleGraha[] = [
  'Sun', 'Moon', 'Mars', 'Mercury', 'Jupiter', 'Venus', 'Saturn',
];

// ── Public API ─────────────────────────────────────────

/**
 * Compute the Varshaphala (Tajik annual horoscope) for a given solar-return
 * year of a native.
 *
 * **Solar return.** A Newton-style refinement around `birthDate +
 * yearAge × 365.25636 days` finds the UTC instant when sidereal Sun =
 * natal sidereal Sun (mod 360°) to within 0.0001° (~9 arcsec). Typical
 * convergence: 3–5 iterations.
 *
 * **Muntha.** `munthaRashi = (natalLagnaRashi + yearAge) mod 12`. Muntha
 * lord = rashi-lord of `munthaRashi`. Muntha house is the whole-sign
 * house (1..12) of `munthaRashi` from the varsha lagna.
 *
 * **Year lord (Varsha Pati).** Selected as the planet with the highest
 * total Shadbala from these 4 candidates (per BPHS Tajik tradition):
 *   1. Varsha lagna lord.
 *   2. Muntha lord.
 *   3. Lord of the Sun's rashi at the varsha instant.
 *   4. Triraashi Pati (3-rashi-trine ruler), per the day/night table.
 *
 * **Sahams.** 27-Saham core set computed in the varsha chart per
 * Neelakantha's *Tajika Neelakanthi*. Day/night flips apply to the 9
 * Sahams flagged in {@link SAHAM_FORMULAS} (Punya, Vidya, Karma, Putra,
 * Roga, Gnati, Bhratri, Matri, Pitri). The extended 50-Saham list is
 * deferred (data-only extension).
 *
 * **Sources.** Neelakantha *Tajika Neelakanthi*; B.V. Raman *Annual
 * Horoscope*; Sanjay Rath *Crux of Vedic Astrology* Tajik appendix;
 * PVR Narasimha Rao Tajik essays.
 *
 * @param natalBirth UTC instant of birth.
 * @param yearAge    1-indexed solar-return year (1 = first solar return,
 *                   typically the day on or near the first birthday). Must
 *                   be a positive integer.
 * @param location   Geographic location for the varsha chart cusps —
 *                   pass the natal location (classical convention) or
 *                   the location at the varsha-time residence.
 * @param options    Birth-chart options (ayanamsa, language, house system).
 *
 * @example
 * ```typescript
 * import { computeVarshaphala } from 'panchang-ts';
 *
 * const varsha = computeVarshaphala(
 *   new Date('1995-08-15T05:30:00Z'),
 *   30,                                              // 30th solar return
 *   { latitude: 28.6139, longitude: 77.2090 },       // New Delhi
 * );
 * varsha.solarReturnInstant;       // ~ 2025-08-15T05:30:Z (within minutes)
 * varsha.varshaLagna.rashi.name;   // ascendant rashi at the return
 * varsha.muntha.lord;              // Muntha lord — graha
 * varsha.yearLord;                 // Varsha Pati (strongest of 4)
 * varsha.sahams.Punya.longitude;   // Punya Saham, [0, 360)
 * varsha.sahams.Punya.house;       // Punya's house from varsha lagna
 * ```
 */
export function computeVarshaphala(
  natalBirth: Date,
  yearAge: number,
  location: GeoLocation,
  options: BirthChartOptions = {},
): VarshaphalaChart {
  validateDate(natalBirth);
  validateLocation(location);
  if (!Number.isInteger(yearAge) || yearAge < 1) {
    throw new PanchangError(
      `Varshaphala yearAge must be a positive integer (1 = first solar return); got ${yearAge}`,
      'INVALID_INPUT',
    );
  }

  const ayanamsaType: AyanamsaType = options.ayanamsa ?? 'lahiri';
  const lang: Language = options.language ?? 'en';

  const natalSun = getSiderealSunLongitude(natalBirth, ayanamsaType);
  const solarReturnInstant = findSolarReturn(natalBirth, yearAge, natalSun, ayanamsaType);

  const varshaChart = computeRashiChart(solarReturnInstant, location, options);
  const isDay = isDayBirth(solarReturnInstant, location);

  const natalLagna = computeLagna(natalBirth, location, ayanamsaType, lang);
  const muntha = buildMuntha(natalLagna.rashi.index, yearAge, varshaChart.lagna.rashi.index, lang);
  const yearLord = pickYearLord(varshaChart, muntha.lord, isDay, solarReturnInstant, location, options);

  const sahams = computeSahams(varshaChart, isDay, lang);

  return {
    solarReturnInstant,
    varshaLagna: varshaChart.lagna,
    muntha,
    yearLord,
    sahams,
    isDayBirth: isDay,
    planets: varshaChart.planets,
    bhava: varshaChart.bhava,
  };
}

// ── Solar return search ────────────────────────────────

/**
 * Locate the UTC instant in year `yearAge` when the sidereal Sun returns
 * to its natal sidereal longitude.
 *
 * Algorithm: Newton-style refinement starting from the calendar
 * approximation `natalBirth + yearAge × SIDEREAL_YEAR_DAYS`. Each step
 * corrects `t` by `−Δλ / SUN_DEG_PER_DAY` days, where Δλ is the wrapped
 * (sidereal-Sun − natal) longitude difference in [-180°, 180°). Converges
 * to 0.0001° (~9 arcsec, ~9 minutes of solar motion / 60 ≈ 9 ms of
 * timing) within 3–5 iterations on cooperative inputs.
 *
 * The 25-iteration cap and 0.0001° tolerance match the plan; in practice
 * the loop exits in a handful of iterations because the Sun's apparent
 * sidereal motion is locally linear over 1-day windows.
 *
 * @internal — exported via `_findSolarReturnForTest` for unit tests.
 */
function findSolarReturn(
  natalBirth: Date,
  yearAge: number,
  natalSun: number,
  ayanamsaType: AyanamsaType,
): Date {
  let t = natalBirth.getTime() + yearAge * SIDEREAL_YEAR_DAYS * 86400_000;

  const TOL_DEG = 0.0001;
  for (let iter = 0; iter < 25; iter++) {
    const lon = getSiderealSunLongitude(new Date(t), ayanamsaType);
    let delta = lon - natalSun;
    delta = ((delta + 540) % 360) - 180; // wrap to [-180, 180)
    if (Math.abs(delta) < TOL_DEG) break;
    t -= (delta / SUN_DEG_PER_DAY) * 86400_000;
  }

  return new Date(Math.round(t));
}

/** @internal */
export const _findSolarReturnForTest = findSolarReturn;

// ── Day / night detection ──────────────────────────────

/**
 * Day-birth flag: true iff the Sun is geometrically above the horizon at
 * the given UTC instant for the given location. Refraction and the Sun's
 * apparent radius are not applied — the classical "day birth" rule uses
 * the geometric center of the Sun.
 *
 * @internal — exported via `_isDayBirthForTest` for unit tests.
 */
function isDayBirth(date: Date, location: GeoLocation): boolean {
  const observer = new Observer(location.latitude, location.longitude, location.elevation ?? 0);
  const time = MakeTime(date);
  const equ = Equator(Body.Sun, time, observer, true, true);
  const hor = Horizon(time, observer, equ.ra, equ.dec, 'normal');
  return hor.altitude > 0;
}

/** @internal */
export const _isDayBirthForTest = isDayBirth;

// ── Muntha ─────────────────────────────────────────────

function buildMuntha(
  natalLagnaRashi: number,
  yearAge: number,
  varshaLagnaRashi: number,
  lang: Language,
): MunthaInfo {
  const munthaRashi = (natalLagnaRashi + yearAge) % 12;
  const lordIdx = RASHI_LORD[munthaRashi]!;
  const lord = VISIBLE_GRAHAS_BY_INDEX[lordIdx]!;
  const house = ((munthaRashi - varshaLagnaRashi + 12) % 12) + 1;
  // lang reserved for future Muntha-rashi name field; kept for API consistency.
  void lang;
  return { rashi: munthaRashi, lord, house };
}

// ── Year lord (Varsha Pati) ────────────────────────────

/**
 * Triraashi Pati — 3-rashi-trine (elemental triplicity) ruler per
 * the standard Tajik table. Element index = `rashi % 4`:
 *
 *   0 → Fire  (Aries / Leo / Sag)
 *   1 → Earth (Taurus / Virgo / Cap)
 *   2 → Air   (Gemini / Libra / Aqua)
 *   3 → Water (Cancer / Scorpio / Pisces)
 *
 * Day rulers: Sun (fire), Venus (earth), Saturn (air), Venus (water).
 * Night rulers: Jupiter (fire), Moon (earth), Mercury (air), Mars (water).
 *
 * Source: B.V. Raman, *Annual Horoscope* Ch. 2.
 */
function triraashiPati(rashi: number, isDay: boolean): VisibleGraha {
  const elem = rashi % 4;
  if (isDay) {
    return (['Sun', 'Venus', 'Saturn', 'Venus'] as const)[elem]!;
  }
  return (['Jupiter', 'Moon', 'Mercury', 'Mars'] as const)[elem]!;
}

/**
 * Year lord (Varsha Pati) — strongest of:
 *   1. Varsha lagna lord
 *   2. Muntha lord
 *   3. Lord of the Sun's rashi at the varsha instant
 *   4. Triraashi Pati of the varsha lagna (day/night rule)
 *
 * "Strongest" = highest total Shadbala from the varsha chart. Tajik bala
 * is itself a Shadbala variant; using Shadbala is the published
 * simplification used by most public Tajik calculators.
 *
 * If two candidates tie on Shadbala, the order above wins (varsha-lagna
 * lord first, then Muntha lord, then Sun's rashi lord, then Triraashi).
 */
function pickYearLord(
  varshaChart: BirthChart,
  munthaLord: VisibleGraha,
  isDay: boolean,
  varshaInstant: Date,
  location: GeoLocation,
  options: BirthChartOptions,
): VisibleGraha {
  const lagnaRashi = varshaChart.lagna.rashi.index;
  const lagnaLord = VISIBLE_GRAHAS_BY_INDEX[RASHI_LORD[lagnaRashi]!]!;

  const sunPlacement = varshaChart.byPlanet.Sun;
  const sunRashiLord = VISIBLE_GRAHAS_BY_INDEX[RASHI_LORD[sunPlacement.rashi.index]!]!;

  const triraashi = triraashiPati(lagnaRashi, isDay);

  // De-dupe in priority order so ties resolve to the earlier candidate.
  const candidates: VisibleGraha[] = [];
  for (const c of [lagnaLord, munthaLord, sunRashiLord, triraashi]) {
    if (!candidates.includes(c)) candidates.push(c);
  }

  const shadbala = computeShadbala(varshaInstant, location, options);

  let best: VisibleGraha = candidates[0]!;
  let bestTotal = shadbala[best].total;
  for (let i = 1; i < candidates.length; i++) {
    const c = candidates[i]!;
    const total = shadbala[c].total;
    if (total > bestTotal) {
      best = c;
      bestTotal = total;
    }
  }
  return best;
}

// ── Sahams ─────────────────────────────────────────────

/**
 * Resolve a {@link SahamOperand} to its sidereal longitude in the varsha
 * chart. `Punya` is special — it must be computed first (the iteration
 * order in {@link SAHAM_FORMULAS} guarantees this) and is then carried in
 * `priorSahams`.
 */
function resolveOperand(
  op: SahamOperand,
  varshaChart: BirthChart,
  priorSahams: Partial<Record<SahamName, number>>,
): number {
  switch (op) {
    case 'Sun':
    case 'Moon':
    case 'Mars':
    case 'Mercury':
    case 'Jupiter':
    case 'Venus':
    case 'Saturn':
      return varshaChart.byPlanet[op].longitude;
    case 'Asc':
      return varshaChart.lagna.siderealLongitude;
    case 'AscLord': {
      const lord = VISIBLE_GRAHAS_BY_INDEX[RASHI_LORD[varshaChart.lagna.rashi.index]!]!;
      return varshaChart.byPlanet[lord].longitude;
    }
    case 'House11Cusp': {
      const lagnaRashi = varshaChart.lagna.rashi.index;
      return ((lagnaRashi + 10) % 12) * 30;
    }
    case 'Punya': {
      const v = priorSahams.Punya;
      if (v === undefined) {
        throw new PanchangError(
          'Saham operand Punya referenced before Punya was computed',
          'SAHAM_DEPENDENCY_ERROR',
        );
      }
      return v;
    }
  }
}

/**
 * Compute one Saham's longitude given its formula. Honors the day/night
 * X/Y swap when `formula.swap` is true and `isDay` is false.
 */
function evaluateSaham(
  formula: SahamFormula,
  isDay: boolean,
  varshaChart: BirthChart,
  priorSahams: Partial<Record<SahamName, number>>,
): number {
  const xOp = formula.swap && !isDay ? formula.y : formula.x;
  const yOp = formula.swap && !isDay ? formula.x : formula.y;
  const x = resolveOperand(xOp, varshaChart, priorSahams);
  const y = resolveOperand(yOp, varshaChart, priorSahams);
  const z = resolveOperand(formula.z, varshaChart, priorSahams);
  return normalize360(x - y + z);
}

/**
 * Compute all 27 Sahams for a varsha chart. Returns a map keyed by Saham
 * name with sidereal longitude, rashi index, localized rashi name, and
 * whole-sign house from the varsha lagna.
 */
function computeSahams(
  varshaChart: BirthChart,
  isDay: boolean,
  lang: Language,
): Record<SahamName, SahamPosition> {
  const longitudes: Partial<Record<SahamName, number>> = {};

  // First pass: longitudes only. Iteration order matters — Punya before
  // Yasas/Mitra/Susha, etc. The formulas table is already in correct order.
  for (const formula of SAHAM_FORMULAS) {
    longitudes[formula.name] = evaluateSaham(formula, isDay, varshaChart, longitudes);
  }

  const lagnaRashi = varshaChart.lagna.rashi.index;
  const out = {} as Record<SahamName, SahamPosition>;
  for (const name of ALL_SAHAM_NAMES) {
    const lon = longitudes[name]!;
    const rashi = Math.floor(lon / 30);
    out[name] = {
      longitude: lon,
      rashi,
      rashiName: resolveMasaName(rashi, lang),
      house: ((rashi - lagnaRashi + 12) % 12) + 1,
    };
  }
  return out;
}

// ── Test-only exports ──────────────────────────────────

/** @internal */
export const _triraashiPatiForTest = triraashiPati;
/** @internal */
export const _evaluateSahamForTest = evaluateSaham;
/** @internal */
export const _resolveOperandForTest = resolveOperand;

File: source/go/internal/jyotish/varshaphala.go
package jyotish

import (
	"math"

	"github.com/ishankgupta95/panchang/source/go/v5/internal/astronomy"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/i18n"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/jsnum"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/utils"
	"github.com/ishankgupta95/panchang/source/go/v5/types"
)

const siderealYearDays float64 = 365.25636

const sunDegPerDay = 360 / siderealYearDays

func ComputeVarshaphala(
	ctx *astronomy.EphemerisCtx,
	natalBirthMs int64,
	yearAge int,
	location types.GeoLocation,
	options BirthChartOptions,
) (VarshaphalaChart, error) {
	if err := utils.ValidateDate(natalBirthMs); err != nil {
		return VarshaphalaChart{}, err
	}
	if err := utils.ValidateLocation(location); err != nil {
		return VarshaphalaChart{}, err
	}
	if yearAge < 1 {
		return VarshaphalaChart{}, types.Codef(types.ErrInvalidInput,
			"Varshaphala yearAge must be a positive integer (1 = first solar return); got %d",
			yearAge)
	}

	ayanamsaType := resolveAyanamsa(options.Ayanamsa)
	lang := resolveLang(options.Language)

	natalSun, err := astronomy.GetSiderealSunLongitude(ctx, natalBirthMs, ayanamsaType)
	if err != nil {
		return VarshaphalaChart{}, err
	}
	solarReturnMs, err := FindSolarReturn(ctx, natalBirthMs, yearAge, natalSun, ayanamsaType)
	if err != nil {
		return VarshaphalaChart{}, err
	}

	varshaChart, err := ComputeRashiChart(ctx, solarReturnMs, location, options)
	if err != nil {
		return VarshaphalaChart{}, err
	}
	isDay := varshaphalaIsDayBirth(ctx, solarReturnMs, location)

	natalLagna, err := ComputeLagna(ctx, natalBirthMs, location, ayanamsaType, lang)
	if err != nil {
		return VarshaphalaChart{}, err
	}
	muntha := buildMuntha(natalLagna.Rashi.Index, yearAge, varshaChart.Lagna.Rashi.Index, lang)
	yearLord, err := pickYearLord(ctx, &varshaChart, muntha.Lord, isDay, solarReturnMs, location, options)
	if err != nil {
		return VarshaphalaChart{}, err
	}

	sahams, err := computeSahams(&varshaChart, isDay, lang)
	if err != nil {
		return VarshaphalaChart{}, err
	}

	return VarshaphalaChart{
		SolarReturnInstant: types.Date(solarReturnMs),
		VarshaLagna:        varshaChart.Lagna,
		Muntha:             muntha,
		YearLord:           yearLord,
		Sahams:             sahams,
		IsDayBirth:         isDay,
		Planets:            varshaChart.Planets,
		Bhava:              varshaChart.Bhava,
	}, nil
}

func FindSolarReturn(
	ctx *astronomy.EphemerisCtx,
	natalBirthMs int64,
	yearAge int,
	natalSun float64,
	ayanamsaType types.AyanamsaType,
) (int64, error) {
	t := float64(natalBirthMs) + float64(float64(yearAge)*siderealYearDays*86400_000)

	const tolDeg = 0.0001
	for iter := 0; iter < 25; iter++ {
		if !jsnum.TimeClip(t) {
			return 0, errInvalidDateInstant
		}
		lon, err := astronomy.GetSiderealSunLongitude(ctx, int64(t), ayanamsaType)
		if err != nil {
			return 0, err
		}
		delta := lon - natalSun
		delta = jsnum.Mod(delta+540, 360) - 180
		if math.Abs(delta) < tolDeg {
			break
		}
		t -= float64((delta / sunDegPerDay) * 86400_000)
	}

	t = jsnum.Round(t)
	if !jsnum.TimeClip(t) {
		return 0, errInvalidDateInstant
	}
	return int64(t), nil
}

func varshaphalaIsDayBirth(ctx *astronomy.EphemerisCtx, ms int64, location types.GeoLocation) bool {
	return astronomy.IsSunAboveHorizon(ctx, ms, location)
}

func buildMuntha(natalLagnaRashi, yearAge, varshaLagnaRashi int, lang types.Language) MunthaInfo {
	munthaRashi := (natalLagnaRashi + yearAge) % 12
	lord := RashiLord[munthaRashi]
	house := (munthaRashi-varshaLagnaRashi+12)%12 + 1
	_ = lang
	return MunthaInfo{Rashi: munthaRashi, Lord: lord, House: house}
}

var (
	triraashiPatiDay = [4]types.VisibleGraha{
		types.VisibleSun, types.VisibleVenus, types.VisibleSaturn, types.VisibleVenus,
	}
	triraashiPatiNight = [4]types.VisibleGraha{
		types.VisibleJupiter, types.VisibleMoon, types.VisibleMercury, types.VisibleMars,
	}
)

func triraashiPati(rashi int, isDay bool) types.VisibleGraha {
	elem := rashi % 4
	if isDay {
		return triraashiPatiDay[elem]
	}
	return triraashiPatiNight[elem]
}

func pickYearLord(
	ctx *astronomy.EphemerisCtx,
	varshaChart *types.BirthChart,
	munthaLord types.VisibleGraha,
	isDay bool,
	varshaInstantMs int64,
	location types.GeoLocation,
	options BirthChartOptions,
) (types.VisibleGraha, error) {
	lagnaRashi := varshaChart.Lagna.Rashi.Index
	lagnaLord := RashiLord[lagnaRashi]

	sunPlacement, ok := varshaChart.ByPlanet.Get(types.GrahaSun)
	if !ok {
		return 0, types.Codef(types.ErrInvalidInput, "varsha chart has no Sun placement")
	}
	sunRashiLord := RashiLord[sunPlacement.Rashi.Index]

	triraashi := triraashiPati(lagnaRashi, isDay)

	candidates := make([]types.VisibleGraha, 0, 4)
	for _, c := range [4]types.VisibleGraha{lagnaLord, munthaLord, sunRashiLord, triraashi} {
		dup := false
		for _, seen := range candidates {
			if seen == c {
				dup = true
			}
		}
		if !dup {
			candidates = append(candidates, c)
		}
	}

	shadbala, err := ComputeShadbala(ctx, varshaInstantMs, location, options)
	if err != nil {
		return 0, err
	}

	best := candidates[0]
	bestBala, _ := shadbala.Get(best)
	bestTotal := bestBala.Total
	for i := 1; i < len(candidates); i++ {
		c := candidates[i]
		b, _ := shadbala.Get(c)
		if b.Total > bestTotal {
			best = c
			bestTotal = b.Total
		}
	}
	return best, nil
}

func resolveOperand(
	op SahamOperand,
	varshaChart *types.BirthChart,
	priorSahams *[SahamNameCount]float64,
	computed *[SahamNameCount]bool,
) (float64, error) {
	switch op {
	case OperandSun, OperandMoon, OperandMars, OperandMercury,
		OperandJupiter, OperandVenus, OperandSaturn:
		p, ok := varshaChart.ByPlanet.Get(types.VisibleGraha(op).Graha())
		if !ok {
			return 0, types.Codef(types.ErrInvalidInput,
				"varsha chart has no %s placement", op)
		}
		return p.Longitude, nil
	case OperandAsc:
		return varshaChart.Lagna.SiderealLongitude, nil
	case OperandAscLord:
		lord := RashiLord[varshaChart.Lagna.Rashi.Index]
		p, ok := varshaChart.ByPlanet.Get(lord.Graha())
		if !ok {
			return 0, types.Codef(types.ErrInvalidInput,
				"varsha chart has no %s placement", lord)
		}
		return p.Longitude, nil
	case OperandHouse11Cusp:
		lagnaRashi := varshaChart.Lagna.Rashi.Index
		return float64(((lagnaRashi + 10) % 12) * 30), nil
	case OperandPunya:
		if !computed[SahamPunya] {
			return 0, types.Codef(types.ErrSahamDependencyError,
				"Saham operand Punya referenced before Punya was computed")
		}
		return priorSahams[SahamPunya], nil
	}
	return 0, types.Codef(types.ErrInvalidInput, "unknown Saham operand %v", op)
}

func evaluateSaham(
	formula SahamFormula,
	isDay bool,
	varshaChart *types.BirthChart,
	priorSahams *[SahamNameCount]float64,
	computed *[SahamNameCount]bool,
) (float64, error) {
	xOp, yOp := formula.X, formula.Y
	if formula.Swap && !isDay {
		xOp, yOp = formula.Y, formula.X
	}
	x, err := resolveOperand(xOp, varshaChart, priorSahams, computed)
	if err != nil {
		return 0, err
	}
	y, err := resolveOperand(yOp, varshaChart, priorSahams, computed)
	if err != nil {
		return 0, err
	}
	z, err := resolveOperand(formula.Z, varshaChart, priorSahams, computed)
	if err != nil {
		return 0, err
	}
	zInArcYtoX := utils.Normalize360(z-y) <= utils.Normalize360(x-y)
	add := 30.0
	if zInArcYtoX {
		add = 0
	}
	return utils.Normalize360(x - y + z + add), nil
}

func computeSahams(varshaChart *types.BirthChart, isDay bool, lang types.Language) (Sahams, error) {
	var longitudes [SahamNameCount]float64
	var computed [SahamNameCount]bool

	for _, formula := range SahamFormulas {
		lon, err := evaluateSaham(formula, isDay, varshaChart, &longitudes, &computed)
		if err != nil {
			return Sahams{}, err
		}
		longitudes[formula.Name] = lon
		computed[formula.Name] = true
	}

	lagnaRashi := varshaChart.Lagna.Rashi.Index
	var out Sahams
	for _, name := range AllSahamNames {
		lon := longitudes[name]
		rashi := int(math.Floor(lon / 30))
		out.Set(name, SahamPosition{
			Longitude: lon,
			Rashi:     rashi,
			RashiName: i18n.ResolveMasaName(rashi, lang),
			House:     (rashi-lagnaRashi+12)%12 + 1,
		})
	}
	return out, nil
}

func TriraashiPatiForTest(rashi int, isDay bool) types.VisibleGraha {
	return triraashiPati(rashi, isDay)
}

func EvaluateSahamForTest(
	formula SahamFormula, isDay bool, varshaChart *types.BirthChart, priorPunya *float64,
) (float64, error) {
	var longitudes [SahamNameCount]float64
	var computed [SahamNameCount]bool
	if priorPunya != nil {
		longitudes[SahamPunya] = *priorPunya
		computed[SahamPunya] = true
	}
	return evaluateSaham(formula, isDay, varshaChart, &longitudes, &computed)
}

func IsDayBirthForTest(ctx *astronomy.EphemerisCtx, ms int64, location types.GeoLocation) bool {
	return varshaphalaIsDayBirth(ctx, ms, location)
}
